### DISCUSSION TITLE
Windows - Debug Build for QPDF?

### DISCUSSION THREAD

**Discussion by chrisba73** [qpdf/qpdf#542] [Q&A]
We have always found a way to just use the binary QPDF Release distributions for Windows, but now, due to other constraints, we have to have a Debug build.  Today, I followed the instructions for the build process [here](https://github.com/qpdf/qpdf/blob/master/README-windows.md), using a fresh MSYS install and a fresh git-clone of QPDF.  I'm using VS 2019.

**Common Setup** - went fine, pretty much the same as the instructions.  Did not try and do `--enable-test-compare-images`.

**External Libraries**  - downloaded prebuilt and placed under top-level qpdf directory - fine.

**Building with MSVC**
**Release:**  
-  ./config-msvc - it failed the first time, but gave a suggestion to run autogen.sh.  Did so, then re-ran ./config-msvc, it worked.  Ran make.  It gave some warnings, but there is a valid build at the end in `build`
- `make check` - <ins>Passes:1736, Failures: 707, Unexpected Passes: 0, Expected Failures: 3</ins>

**Baseline:**
I didn't know what normal test results were, so I copied out `build` to `build_release`, deleted .lib and .dll in `build`, then copied over the binary files from the qpdf-10.3.2-bin-msvc64.zip release into `build`.
- `make check` - <ins>Passes:1739, Failures: 704, Unexpected Passes: 0, Expected Failures: 3</ins>
- [then, renamed `build` to `baseline`].

**Debug:**
- I made a Debug build.  The instructions provided almost worked for me, but I had to make changes from  -MD to -MDd in msvc.mk and also 1 change in autoconf.ml, from `-lmsvcrt` to `-lmsvcrtd`, to resolve link errors.  Then, make finished and gave binaries in `build`, but there were many more warnings (many about PDBs, but also LNK4098 - defaultlib MSVCRT conflicts with use of other libs; use /NODEFAULTLIB:library, on all of the links).  I'm still using the prebuilt binaries here for crypto.lib, jpeg.lib, ssl.lib, and z.lib.
- `make check` - <ins>failed most tests (I see only 6 passes)</ins>, and also produced some crashes on some.  Finally erred out at qpdf 498, 'Died at qpdf.test line 1423.  invalid results from subsidiary test driver'.
- I added `-NODEFAULTLIB:MSVCRT` under `define makelib` and `define makebin`.  The LNK4098 warnings went away, but the test results are the same.

So, I'm still working through things, maybe it's a simple problem to get the Debug Pass/Fail performance to the level of the binary releases or the build(s) that I'm doing.  

**Questions**
- What is the expectation for the current test suite, for Passes/Failures with `make check`?  I can send more information on what passed and what didn't if needed, but, for example, all of the Form XObject tests (129-162) fail for `baseline` and `build_release`.
- Any suggestions? (Do you think my issue might be that I need to build the external dependencies as Debug as well?)
- Any plans in the future for CMAKE support?

**Comment by jberkenbilt** [comment]
I'm afraid my experience on Windows is very limited...basically I know enough to get qpdf building and working well on that platform, but I don't really have any experience with Windows development, and the work I did on porting to Windows was done over 10 years ago. Based on my memory, you would need debugging versions of the third party libraries, but I could be mistaken. Maybe someone with more Windows development experience could chime in on this.

To answer your other questions, the expectation on tests is that there are three expected failures, and everything else should pass. I don't know what I would do with information about what passed and what failed since I don't have a good Windows development environment set up anymore. I basically have a Windows VM with the same toolchain installed as in the github actions runner that builds qpdf in CI. This way I can debug Windows build failures. It's been a very long time since I've had to actually dig into any Windows-specific issue. If you discover something I can change about the current build to make it more able to support a Windows debug build, I'd welcome a patch or pull request.

Regarding cmake, I have considered it and would be open to a pull request, or even just an issue requesting cmake support, which I could tag as an enhancement. When I originally created qpdf in 2002, it was part of a large system whose life started prior to cmake's existence. At that time, I had worked on an open source build system called abuild, which I eventually dropped as it had been clearly overtaken by other tools. If I were starting today, it is very likely that I would use cmake, and perhaps it's time to consider migrating to it. I suspect that it would simplify situations like yours and probably reduce other problems that have happened with the build, such as one issue involving packaging rpms.

Sorry I can't be of more immediate help on this.

**Reply by chrisba73** [reply]
OK, thanks for the feedback.  If I can make progress on a Windows build, I'll be sure to let you know (or if others have, that would be great to know too)!  

If all of the tests should pass - I'm not sure what then means then, since the packaged 10.3.2 Windows binaries don't seem to do that on my computer, unless there's something wrong with the way I'm running the `make check` test against those binaries.  Not sure, maybe I should address that first.

Would you mind spot-checking one test on your Windows VM (if it's easy to do so)?  It might help decide which direction to proceed.  Picking one that looks easy - for test "qpdf 129", the first Form XObject test [for fxo-red.pdf] - the qtest.log file for the 10.3.2 binaries shows:
```
--> BEGIN EXPECTED OUTPUT <--
test 55 done
--> END EXPECTED OUTPUT <--
--> BEGIN ACTUAL OUTPUT <--
WARNING: fxo-red.pdf: file is damaged
WARNING: fxo-red.pdf (offset 7568): xref not found
WARNING: fxo-red.pdf: Attempting to reconstruct cross-reference table
WARNING: fxo-red.pdf (object 20 0, offset 4424): expected endstream
...
```
When I run qpdf against the fxo-red.pdf file (on Windows) outside of the qtest framework, I get the same sort of warnings:
```
WARNING: filebuffer: file is damaged
WARNING: filebuffer (offset 7568): xref not found
WARNING: filebuffer: Attempting to reconstruct cross-reference table
```

**Reply by jberkenbilt** [reply]
I'm wondering whether you ran `git config --global core.autocrlf input` before your clone. Git will guess wrong about ASCII vs. binary on Windows for lots of test files in the qpdf test suite because PDF files are a mixture of ASCII and binary.

**Reply by jberkenbilt** [reply]
You don't have to use `--global` -- just make sure `core.autocrlf` is `input` for this repository.

**Reply by chrisba73** [reply]
I did not.  Thanks, that could explain a lot of things!  OK, I'll try a fresh run this afternoon.

**Reply by jberkenbilt** [reply]
I made a note to myself to update the README files to make sure to mention this. I also made a note to myself to look into CMake. I looked at it very briefly in 2007 and haven't touched it since. :-)

**Reply by chrisba73** [reply]
That made a huge difference!  And noticeably, in the setup process: /config-msvc did not fail, there was no need to run autogen.sh.

**Release Build**: 2443/2446 qpdf tests passed; 5251 fuzz tests passed; tests after that were passing until pdf-mod-info 14 caused a glitch with Norton (now whitelisted; will go back to this later).

**Debug Build**:  4 tests failed: 364 to 367 (NNtree).   [This is using -NODEFAULTLIB:MSVCRT, but the external-libs libraries as-is, not rebuilding them.]

<ins>Test 364</ins>: In the Visual Studio debugger, the issue appears to be initiated from check_split1(15) in test-case 74, and then at line 486 of NNTree.c :
```
    if (! is_root)
    {
        QTC::TC("qpdf", "NNTree split parent");
        auto next = parent->node;
        resetLimits(next, parent);
        --parent;     // <===== Problem here
        split(next, parent);
    }
```
where operator--() invokes:
`_STL_VERIFY(_New_ptr != _Mycont->_Myhead, "cannot decrement begin list iterator");`

<ins>Test 366</ins>:   erase3.remove(310); leads to the following in NNtree.cc, lines 603-608:
```
    while (! done)
    {
        auto element = lastPathElement();
        auto parent = element;
        --parent;    // <===== Problem here
        auto kids = element->node.getKey("/Kids");
    ...
```
where it again complains about the --parent in the same way as Test 364.

I don't know if you have any thoughts on whether these are 'common code areas' or not, but I may proceed ahead with these Debug libs/DLLs, with the thought that they may be good-enough for our debugging needs for now.

**Reply by jberkenbilt** [reply]
Sounds like you've made great progress. Thanks for keeping this updated. Those are not common code areas. There's probably nothing in the library that currently calls that, though I wouldn't swear to it. I just added that stuff recently. I'll have to look into that -- you may have uncovered a real bug, or at the very least, you may have uncovered an assumption I was making about STL lists that happens to be true with gcc and clang but not Visual Studio. I should probably adjust the code so that it doesn't try to decrement a begin list iterator. I vaguely remember having determined that decrementing a list begin iterator brought you back to the end and relying on that behavior, but I can't remember whether I kept that code or not.

In any case, for any kind of normal use of the library, you are probably safe to proceed with those failures. I will again make a note to myself to create an actual issue from this because I think all tests should pass with a debug build too.

**Comment by jberkenbilt** [comment]
By the way, I spent a little time reading through some CMake documentation to try to catch myself up a little on what's changed since I last looked at it in 2007. I think qpdf would probably benefit from being converted over to cmake since the current build system is very much home-grown -- it uses autoconf but not automake. I also believe it would be a pretty big project. My configure.ac is almost 1,000 lines of code, and my build does lots of custom stuff. The qpdf codebase is almost 20 years old, so it's probably due for modernization in this particular area. No promises. It would be a big job, and there are other things competing for my attention.

**Reply by chrisba73** [reply]
I agree, I think it would be a very large effort to move qpdf over to CMAKE.  I'm a beginner with CMAKE, but I find it tedious to construct and very slow to debug.  And, one thing that I like is to be able to configure multiple platform releases at a time (instead of command-line cmake debug, cmake release, cmake debug-with-release-information), so that you end up with a single SLN file for Visual Studio, and that seems challenging for various reasons, especially when working with external libraries.

The answer [here](https://stackoverflow.com/questions/7132862/how-do-i-convert-an-autotools-project-to-a-cmake-project) is old, but it makes it appear that there are no shortcuts in converting to CMAKE.

Maybe using CMAKE just for building certain platforms, like Windows, is an intermediate step (especially since external libraries seem like they can be integrated easily).  One thing that may help the automake conversion is that it seems like platform-type checks are often one-liners in CMAKE; see an example [here](https://github.com/OpenChemistry/openchemistry/blob/master/cmake/libarchive.CMakeLists.txt), for instance and the wiki [here](https://gitlab.kitware.com/cmake/community/-/wikis/doc/tutorials/How-To-Write-Platform-Checks).

**Comment by jberkenbilt** [comment]
qpdf 11 will use cmake. Conversion of qpdf to cmake has been in main for a while now but is not yet in any released version.
